### ISSUE TITLE
[rush] [feature] support for git submodules (they currently break rush build algo)

### ISSUE THREAD

**Issue by trusktr** [microsoft/rushstack#1711]
EDIT, this bug converted into a feature request.

**Is this a feature or a bug?**

- [x] Feature
- [ ] Bug

**Please describe the actual behavior.**

When I start clean, `rush build` builds all my projects that are listed in rush.json.

After I edit a file in one project, then when I run `rush build` again it skips the project and does not build it.

**What is the expected behavior?**

I was expecting rush to build the project because I modified it.

**workaround**

for now I have to use `rush rebuild --to package-name` to force it to rebuild the package.

**If this is a bug, please provide the tool version, Node.js version, and OS.**

* **Tool:** rush
* **Tool Version:** 5.18.0
* **Node Version:** 13.6.0
* **OS:** NixOS

**Comment by iclanton** [comment]
Is the file you edited ignored by `.gitignore`? Is it listed in `<project>/.rush/temp/package-deps_build.json`?

**Comment by trusktr** [comment]
The file is not ignored by `.gitignore`, and in `<project>/.rush/temp/package-deps_build.json` I see:

```json
{
  "files": {},
  "arguments": "tsc -p ./tsconfig.json "
}
```

**Comment by trusktr** [comment]
Ooooooh. I think I see the issue:

My `<projects>` are git sub modules. The umbrella repo does not see what files changed.

Is there an alternative way to tell rush which files to check for changes?

**Comment by trusktr** [comment]
I'd like to keep each project as a sub-module, because then they carry their own history which is not intertwined with other packages.

Having them as separate repos also allows them to be installable from git, not just from NPM. For example, a consuming project's `package.json` can have:

```json
  "dependencies": {
    "the-package": "github-username/repo-name#some-branch"
  }
```

which makes it easy to temporarily rely on git refs from forks, etc.

**Comment by apostolisms** [comment]
This is something we haven't tested. This should become a feature request but we will need further design and a better understanding of what it means to support this scenario. 

This is something we might want to think about including in an upcoming major release.

**Comment by trusktr** [comment]
I haven't looked at the source yet, but it seems that at the point that Rush detects changes, instead of running `getFiles(exec('git diff ...'))` (or similar pseudo code) it could run `projects.reduce(p => getFiles(exec('cd '+p.path+' && git diff ...')), [])` to get the list of modified files.

**Comment by NullMDR** [comment]
@apostolisms Any update on this item? Currently we have a lot of git repos and we want to migrate them smoothly with submodule, but currently we are not sure about rush support for submodule. Rush build failed with

`
Error calculating the state of the repo. (inner error: Error: git hash-object exited with status 128: fatal: Unable to hash /home/htc/openapi-alps/@azure/rest-api-specs-scripts
). Continuing without diffing files.
`

**Comment by octogonz** [comment]
One of the main benefits of a monorepo is the "you broke it, you fix it" property:

*A PR cannot merge unless it successfully passes the tests for all downstream projects.*

Without this, as you scale up, when one team makes a change to an upstream project, they can break another team. That team becomes a downstream "victim" because they don't encounter the problem until later, when they need to upgrade to the latest version.  Normally this problem arises via NPM publishing (see [this article](https://rushjs.io/pages/intro/why_mono/) for a basic sketch).  But in the case if Git submodules it will happen when you update the submodule hash.  The monorepo avoids this by pushing responsibility for breaks back onto the person who wants to make the change, avoiding downstream victims.

With submodules, you could work around this problem by using CI for each submodule repo, i.e. blocking PRs from being merged unless the main repo first completes a successful build.  But this reproduces the exact same pros/cons of a monorepo, just in a more awkward form.  So technically submodules don't seem to bring much benefit.

One benefit of submodules is that they would be presented better via GitHub, because GitHub's design (today) does not consider monorepos very much. Today, GitHub's UI assumes each repo is a single project.  GitHub should really fix that. Other systems have better support for monorepos.

If you want to proceed anyway, it sounds like certain Rush operations assume that there is only one repo -- maybe we can relax that.  For the incremental build issue you pointed out, the `package-deps-hash` library leverages Git as an optimization to quickly calculate file hashes. We could probably generalize that to work with submodules.  We would certainly accept PRs to make these improvements. (However I'm not very interested in investing in that myself, as I don't see how submodules can scale.  We try to focus on battle-tested strategies that will scale.)  But if someone else wants to implement this, go for it.

**Comment by octogonz** [comment]
> I haven't looked at the source yet, but it seems that at the point that Rush detects changes, instead of running `getFiles(exec('git diff ...'))` (or similar pseudo code) it could run `projects.reduce((a, p) => a.concat(getFiles(exec('cd '+p.path+' && git diff ...'))), [])` to get the list of modified files.

This would be slower. If we need a workaround like this, it may need to be behind an "opt-in" setting that enables support for submodules.

The `rush change` command may also have trouble, since it tries to determine what has changed since the base branch of your PR, which also involves some Git assumptions. Probably also solvable, though.

**Comment by octogonz** [comment]
> Currently we have a lot of git repos and we want to migrate them smoothly to mono repo with submodule, 

Why not create a new clean monorepo, and then merge the projects into it one by one? IMO this will go a lot smoother than trying to maintain two models side by side.

**Comment by NullMDR** [comment]
@octogonz Thank you for your reply! Our problem is a little bit complicated: we have several public repos and several private repos, and we still want to maintain them with monorepo. That means we need to synchronize between public repo and private monorepo. Previously I'm considering git submodules, but after some research I found git subtree. We can push/pull from a git subtree which meet our needs, however I'm not sure about limitation of git subtree with rush. Do you have any comment on our situation? Thanks in advance!

**Comment by NullMDR** [comment]
@octogonz Oh yes thanks a lot for your explanation! Seems like I need to dig into git subtree's internal to evaluate but in general it should be fine. We will report our experience after migrating to this solution.

**Comment by GongT** [comment]
submodule also break cache: (/docs is a submodule)
```
Error calculating the state of the repo. (inner error: Error: git hash-object exited with status 128: fatal: Unable to hash docs
). Continuing without diffing files.
```

**Comment by dystopiandev** [comment]
Just migrated from lerna to rush. This is my only problem so far.

**Comment by octogonz** [comment]
> Just migrated from lerna to rush. This is my only problem so far.

We would like to support submodules, and would happily accept a community PR. However I suspect the official Rush maintainers are unlikely to implement it, because as far as I know, nobody is wanting submodules in any of our monorepos.  Every time someone suggests it, it seems like they always end up preferring a different solution.

**Comment by towry** [comment]
Is there some alternative tools that support git submodule ??

**Comment by octogonz** [comment]
> Is there some alternative tools that support git submodule ??

Looks like some support was implemented in https://github.com/microsoft/rushstack/pull/233

And @chengcyber has created PR https://github.com/microsoft/rushstack/pull/3749 that maybe will fix some other problems.

**Comment by octogonz** [comment]
Thanks @chengcyber for fixing this in PR #3749!

🚀 Released with Rush 5.83.3
